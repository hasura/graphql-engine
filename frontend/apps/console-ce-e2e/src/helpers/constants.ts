export const ADMIN_SECRET_HEADER_KEY = 'x-hasura-admin-secret';
