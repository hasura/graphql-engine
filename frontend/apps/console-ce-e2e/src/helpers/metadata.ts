export const replaceMetadata = (newMetadata: Record<string, any>) => {
  const postBody = { type: 'replace_metadata', args: newMetadata };
  cy.request('POST', 'http://localhost:8080/v1/metadata', postBody).then(
    response => {
      expect(response.body).to.have.property('message', 'success'); // true
    }
  );
};

export const resetMetadata = () => {
  const postBody = { type: 'clear_metadata', args: {} };
  cy.request('POST', 'http://localhost:8080/v1/metadata', postBody).then(
    response => {
      expect(response.body).to.have.property('message', 'success'); // true
    }
  );
};
