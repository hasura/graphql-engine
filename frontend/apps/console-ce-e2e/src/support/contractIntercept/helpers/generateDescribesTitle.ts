/**
 * Return an array containing all the "describe"s title.
 *
 * Ex. return ['Describe 1', 'Describe 2'] in this case
 * describe('Describe 1', () => {
 *   describe('Describe 2', () => {
 *     it('Test', () => {})
 *   })
 * })
 */
export function generateDescribesTitle(
  parent?: Mocha.Suite | undefined
): string[] {
  if (!parent) return [];

  let parentNames: string[] = [];
  // parents are recursive
  if (parent.parent) {
    parentNames = parentNames.concat(generateDescribesTitle(parent.parent));
  }

  if (parent.title !== '') {
    parentNames.push(parent.title);
  }

  return parentNames;
}
