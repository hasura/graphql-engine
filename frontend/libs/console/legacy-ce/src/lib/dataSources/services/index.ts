import { postgres } from './postgresql';
import { mysql } from './mysql';
import { mssql } from './mssql';
import { bigquery } from './bigquery';
import { citus } from './citus';
import { cockroach } from './cockroach';

export const services = {
  postgres,
  mysql,
  mssql,
  bigquery,
  citus,
  cockroach,
};
