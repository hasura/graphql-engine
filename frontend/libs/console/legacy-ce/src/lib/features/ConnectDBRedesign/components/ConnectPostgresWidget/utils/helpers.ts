import { isProConsole } from '../../../../../utils';
import { PostgresConnectionInfoSchema } from '../schema';

export const getDatabaseConnectionDisplayName = (
  databaseUrl: PostgresConnectionInfoSchema['databaseUrl']
) => {
  return databaseUrl.connectionType === 'databaseUrl'
    ? databaseUrl.url
    : databaseUrl.connectionType === 'envVar'
    ? databaseUrl.envVar
    : getDatabaseUrlFromConnectionParams({
        username: databaseUrl.username,
        password: databaseUrl.password,
        database: databaseUrl.database,
        host: databaseUrl.host,
        port: databaseUrl.port,
      });
};

export const getDatabaseUrlFromConnectionParams = ({
  username,
  password,
  database,
  host,
  port,
}: {
  username: string;
  password?: string;
  database: string;
  host: string;
  port: number;
}) => `postgresql://${username}.${password ?? ''}@${host}:${port}/${database}`;

const isFalseyValue = (v: any) =>
  Number.isNaN(v) ||
  v === undefined ||
  v === null ||
  (typeof v === 'object' && Object.keys(v).length === 0) ||
  (typeof v === 'string' && v.length === 0);

export const cleanEmpty = (obj: Record<string, any>): any => {
  if (Array.isArray(obj)) {
    return obj
      .map(v => (v && typeof v === 'object' ? cleanEmpty(v) : v))
      .filter(v => !isFalseyValue(v));
  } else {
    return Object.entries(obj)
      .map(([k, v]) => [k, v && typeof v === 'object' ? cleanEmpty(v) : v])
      .reduce(
        (a, [k, v]) => (isFalseyValue(v) ? a : ((a as any)[k] = v), a),
        {}
      );
  }
};

export const areSSLSettingsEnabled = () => {
  return isProConsole(window.__env);
};

export const areReadReplicasEnabled = () => {
  return isProConsole(window.__env) || true;
};
