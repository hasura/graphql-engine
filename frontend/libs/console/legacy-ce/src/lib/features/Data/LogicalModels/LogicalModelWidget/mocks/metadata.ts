export const metadata = {
  resource_version: 5,
  metadata: {
    version: 3,
    sources: [
      {
        name: 'chinook',
        kind: 'postgres',
        tables: [
          {
            table: {
              name: 'Album',
              schema: 'public',
            },
          },
        ],
        logical_models: [
          {
            fields: [
              {
                name: 'a',
                nullable: false,
                type: 'text',
              },
            ],
            name: 'a',
          },
          {
            fields: [],
            name: 'as',
          },
        ],
        configuration: {
          connection_info: {
            database_url: 'postgres://postgres:pass@postgres:5432/chinook',
            isolation_level: 'read-committed',
            use_prepared_statements: false,
          },
        },
        customization: {
          root_fields: {
            namespace: 'postgres',
          },
        },
      },
      {
        name: 'bikes',
        kind: 'mssql',
        tables: [],
        stored_procedures: [
          {
            arguments: {
              test: {
                nullable: true,
                type: 'varchar',
              },
            },
            configuration: {
              exposed_as: 'query',
            },
            returns: 'logical_model_1',
            stored_procedure: {
              name: 'stored_procedure_1',
              schema: 'dbo',
            },
          },
          {
            arguments: {
              name: {
                nullable: true,
                type: 'varchar',
              },
            },
            configuration: {
              exposed_as: 'query',
            },
            returns: 'logical_model_2',
            stored_procedure: {
              name: 'stored_procedure_2',
              schema: 'dbo',
            },
          },
        ],
        logical_models: [
          {
            fields: [
              {
                name: 'TABLE_QUALIFIER',
                nullable: false,
                type: 'sysname',
              },
              {
                name: 'TABLE_OWNER',
                nullable: false,
                type: 'sysname',
              },
              {
                name: 'TABLE_NAME',
                nullable: false,
                type: 'sysname',
              },
              {
                name: 'TABLE_TYPE',
                nullable: false,
                type: 'varchar',
              },
              {
                name: 'REMARKS',
                nullable: false,
                type: 'varchar',
              },
            ],
            name: 'logical_model_1',
          },
          {
            fields: [
              {
                name: 'TABLE_QUALIFIER',
                nullable: false,
                type: 'sysname',
              },
              {
                name: 'TABLE_OWNER',
                nullable: false,
                type: 'sysname',
              },
              {
                name: 'TABLE_NAME',
                nullable: false,
                type: 'sysname',
              },
              {
                name: 'TABLE_TYPE',
                nullable: false,
                type: 'varchar',
              },
              {
                name: 'REMARKS',
                nullable: false,
                type: 'varchar',
              },
            ],
            name: 'logical_model_2',
          },
        ],
        configuration: {
          connection_info: {
            connection_string:
              'DRIVER={ODBC Driver 17 for SQL Server};SERVER=host.docker.internal;DATABASE=bikes;Uid=SA;Pwd=reallyStrongPwd123',
            pool_settings: {
              idle_timeout: 5,
              max_connections: null,
              total_max_connections: null,
            },
          },
        },
      },
    ],
    backend_configs: {
      dataconnector: {
        mysqlgdc: {
          uri: 'http://data-connector-agent:8081/api/v1/mysql',
        },
        sqlite: {
          uri: 'http://sqlite:8100',
        },
      },
    },
  },
};
