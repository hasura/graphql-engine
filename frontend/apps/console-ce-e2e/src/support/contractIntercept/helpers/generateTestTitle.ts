import type { TestTitle } from '../types';
import { generateDescribesTitle } from './generateDescribesTitle';

/**
 * Return a string containing all the "describe"s titles and the test title concatenated.
 *
 * Ex. return 'Describe 1 - Describe 2 - Test' in this case
 * describe('Describe 1', () => {
 *   describe('Describe 2', () => {
 *     it('Test', () => {})
 *   })
 * })
 */
export function generateTestTitle(thisTest: Mocha.Context): TestTitle {
  const describesTitle = generateDescribesTitle(thisTest.parent);
  const testTitle = thisTest.title;

  return [...describesTitle, testTitle].join(' - ');
}
