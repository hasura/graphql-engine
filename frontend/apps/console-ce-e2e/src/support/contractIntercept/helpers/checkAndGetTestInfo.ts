import { generateTestTitle } from './generateTestTitle';
import { throwIfCalledInTestHooks } from './throwIfCalledInTestHooks';
import { throwIfCalledInsideArrowFunction } from './throwIfCalledInsideArrowFunction';

/**
 * Perform pre-flight checks and return the most important test info.
 */
export function checkAndGetTestInfo(thisTest: Mocha.Context) {
  const testTitle = generateTestTitle(thisTest); // ex. 'Describe 1 - Describe 2 - Describe 3 - Test title'
  const testPath = thisTest?.invocationDetails?.relativeFile;

  throwIfCalledInTestHooks(thisTest.title);
  throwIfCalledInsideArrowFunction(thisTest.title);

  // TS-only check, it should never happen at runtime
  if (typeof testPath !== 'string') {
    throw new Error(`No test Path available for ${testTitle}`);
  }

  return {
    testPath,
    testTitle,
  };
}
