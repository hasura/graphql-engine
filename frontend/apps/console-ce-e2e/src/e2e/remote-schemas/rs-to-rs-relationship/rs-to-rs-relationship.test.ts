import { replaceMetadata } from '../helpers/metadata';
import { postgres } from '../../data/manage-database/postgres.spec';
import { HasuraMetadataV3 } from '@hasura/console-legacy-ce';
import { readMetadata } from '../../actions/withTransform/utils/services/readMetadata';

describe('check if remote schema to remote schema relationships are created properly', () => {
  before(() => {
    // create a table called destination_table
    postgres.helpers.createTable('destination_table');

    // load stuff into the metadata
    replaceMetadata({
      version: 3,
      sources: [
        {
          name: 'default',
          kind: 'postgres',
          tables: [
            {
              table: {
                schema: 'public',
                name: 'destination_table',
              },
            },
          ],
          configuration: {
            connection_info: {
              use_prepared_statements: true,
              database_url: {
                from_env: 'HASURA_GRAPHQL_DATABASE_URL',
              },
              isolation_level: 'read-committed',
              pool_settings: {
                connection_lifetime: 600,
                retries: 1,
                idle_timeout: 180,
                max_connections: 50,
              },
            },
          },
        },
      ],
      remote_schemas: [
        {
          name: 'source_rs',
          definition: {
            url: 'https://graphql-pokemon2.vercel.app/',
            timeout_seconds: 60,
          },
          comment: '',
        },
        {
          name: 'ref_rs',
          definition: {
            url: 'https://countries.trevorblades.com/',
            timeout_seconds: 60,
          },
          comment: '',
        },
      ],
    });
  });

  it('verify creating a new rs-to-db relationship', () => {
    cy.visit('/remote-schemas/manage/source_rs/relationships');
    cy.findByText('Add a new relationship').click();
    cy.findByText('Remote Schema').click();
    cy.get('[name=name]').type('RelationshipName');
    cy.get('[name=rsSourceType]').select('Pokemon');
    cy.get('[name=referenceRemoteSchema]').select('ref_rs');
    cy.get('.ant-tree-switcher').click();
    cy.get('.ant-tree-switcher_close').first().click();
    cy.findByRole('button', { name: 'Add Relationship' }).click();

    cy.get('[data-test=remote-schema-relationships-table').should('exist');
    cy.get('[data-test=remote-schema-relationships-table')
      .find('tr')
      .should('have.length', 2);
    cy.get('[data-test=remote-schema-relationships-table').contains(
      'td',
      'RelationshipName'
    );
    readMetadata().then((md: { body: HasuraMetadataV3 }) => {
      cy.wrap(
        md.body?.remote_schemas?.find(rs => rs?.name === 'source_rs')
          ?.remote_relationships
      ).toMatchSnapshot({ name: 'rs-to-rs-relationship' });
    });

    // delete rs-to-rs relationship
    cy.visit('/remote-schemas/manage/source_rs/relationships', {
      timeout: 10000,
      onBeforeLoad(win) {
        cy.stub(win, 'prompt').returns('RelationshipName');
      },
    });
    cy.findByRole('button', { name: 'Remove' }).click();

    //  delete both remote schemas
    cy.visit('/remote-schemas/manage/source_rs/modify', {
      timeout: 10000,
      onBeforeLoad(win) {
        cy.stub(win, 'prompt').returns('source_rs');
      },
    });
    cy.get('[data-test=remote-schema-edit-delete-btn]').click();
    cy.visit('/remote-schemas/manage/ref_rs/modify', {
      timeout: 10000,
      onBeforeLoad(win) {
        cy.stub(win, 'prompt').returns('ref_rs');
      },
    });
    cy.findByRole('button', { name: 'Delete' }).click();

    // delete the table
    postgres.helpers.deleteTable('destination_table');
  });
});
