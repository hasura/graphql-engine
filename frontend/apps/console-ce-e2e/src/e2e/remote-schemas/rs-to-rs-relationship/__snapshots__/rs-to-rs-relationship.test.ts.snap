exports[`check if remote schema to remote schema relationships are created properly > verify creating a new rs-to-db relationship #0`] =
[
  {
    "relationships": [
      {
        "definition": {
          "to_remote_schema": {
            "lhs_fields": [],
            "remote_field": {
              "continent": {
                "arguments": {}
              }
            },
            "remote_schema": "ref_rs"
          }
        },
        "name": "RelationshipName"
      }
    ],
    "type_name": "Pokemon"
  }
];
